File: easy/ruby/best_time_to_buy_and_sell_stock.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
  buy_price = prices[0]
  max_profit = 0
  prices.each do |price|
    profit = price - buy_price
    if max_profit < profit
      max_profit = profit
    elsif price < buy_price
      buy_price = price
    end
  end

  max_profit
end

File: easy/ts/best_time_to_buy_and_sell_stock.ts
function maxProfit(prices: number[]): number {
  let maxProfit = 0;
  let buyDay = prices[0];
  for (let i = 0; i < prices.length; i++) {
    const profit = prices[i] - buyDay
    if (maxProfit < profit) {
      maxProfit = profit;
    }
    if (prices[i] < buyDay) {
      buyDay = prices[i];
    }
  }

  return maxProfit
};